internal/controller/servermaintenance_controller.go

```diff
@@ -0,0 +1,484 @@
+		return ctrl.Result{}, err
+	}
+	config, err := r.createServerBootConfiguration(ctx, serverMaintenance, server)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if config == nil {
+		if err := r.setAndPatchServerPowerState(ctx, server, serverMaintenance); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	if config.Status.State == metalv1alpha1.ServerBootConfigurationStatePending || config.Status.State == "" {
+		log.V(1).Info("Server boot configuration is pending", "Server", server.Name)
+		return ctrl.Result{}, nil
+	}
+	if config.Status.State == metalv1alpha1.ServerBootConfigurationStateError {
+		if modified, err := r.patchMaintenanceState(ctx, serverMaintenance, metalv1alpha1.ServerMaintenanceStateFailed); err != nil || modified {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	if config.Status.State == metalv1alpha1.ServerBootConfigurationStateReady {
+		log.V(1).Info("Server maintenance boot configuration is ready", "Server", server.Name)
+		if err := r.setAndPatchServerPowerState(ctx, server, serverMaintenance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ServerMaintenanceReconciler) createServerBootConfiguration(ctx context.Context, maintenance *metalv1alpha1.ServerMaintenance, server *metalv1alpha1.Server) (*metalv1alpha1.ServerBootConfiguration, error) {
+	log := ctrl.LoggerFrom(ctx)
+	if maintenance.Spec.ServerBootConfigurationTemplate == nil {
+		log.V(1).Info("No ServerBootConfigurationTemplate specified")
+		return nil, nil
+	}
+
+	log.V(1).Info("Creating/Patching server maintenance boot configuration", "Server", server.Name)
+	config := &metalv1alpha1.ServerBootConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      maintenance.Name,
+			Namespace: maintenance.Namespace,
+		},
+	}
+	opResult, err := controllerutil.CreateOrPatch(ctx, r.Client, config, func() error {
+		config.Spec = maintenance.Spec.ServerBootConfigurationTemplate.Spec
+		return controllerutil.SetControllerReference(maintenance, config, r.Scheme)
+	})
+	if err != nil {
+		return config, fmt.Errorf("failed to create server boot configuration: %w", err)
+	}
+	log.V(1).Info("Created or patched Config", "Config", config.Name, "Operation", opResult)
+	serverBase := server.DeepCopy()
+	server.Spec.MaintenanceBootConfigurationRef = &v1.ObjectReference{
+		Namespace:  config.Namespace,
+		Name:       config.Name,
+		UID:        config.UID,
+		APIVersion: "metal.ironcore.dev/v1alpha1",
+		Kind:       "ServerBootConfiguration",
+	}
+	if err := r.Patch(ctx, server, client.MergeFrom(serverBase)); err != nil {
+		return config, fmt.Errorf("failed to patch server maintenance boot configuration ref: %w", err)
+	}
+	return config, nil
+}
+
+func (r *ServerMaintenanceReconciler) setAndPatchServerPowerState(ctx context.Context, server *metalv1alpha1.Server, maintenance *metalv1alpha1.ServerMaintenance) error {
+	log := ctrl.LoggerFrom(ctx)
```

Comment:
Can we pass the log from the outside to this method.

Reply:
yes